### Analysis pipeline for the identification of DNA differentially methylated sites between two groups for MeDIP-Seq (Illumina paired-end reads) with multiple biological replicates in each condition/group.

# Steps of DMSA (Differentially Methylated Sites Analyzer)

## Please define PATH of your folders before you start this analysis pipeline.
# PATH to the reference genome
PATH_To_Ref="/clusterdata/hiseq_apps/resources/freeze001/mm9"

# PATH to the fastq file
PATH_To_Fastq="/clusterdata/uqqzhao/QBI_projects/MeDIP_Danay/fastq"

# PATH for saving the alignment results generated by bwa
PATH_To_AlignmentResults="/clusterdata/uqqzhao/QBI_projects/MeDIP_Danay/BWA0.6.2"

# PATH for saving the peaks results generated by MACS
PATH_To_peaks="/clusterdata/uqqzhao/QBI_projects/MeDIP_Danay/MACS_peaks"

# PATH for saving the result files (inlcuding all peak positons, differentially methylated peaks, etc.)
PATH_To_results="/clusterdata/uqqzhao/QBI_projects/MeDIP_Danay/results"



########################### No need to change for the below context ################################
# PATH for saving the sam and bam files
PATH_To_BAM="$PATH_To_AlignmentResults/BAM";

# PATH for saving the statistical summary files for mapping results
PATH_To_stats="$PATH_To_AlignmentResults/stats";

# PATH for saving the raw count data and normalised count data
PATH_To_counts="$PATH_To_results/counts";


if [ ! -e $PATH_To_AlignmentResults ]; then mkdir $PATH_To_AlignmentResults ; fi
if [ ! -e $PATH_To_AlignmentResults/raw ]; then mkdir $PATH_To_AlignmentResults/raw ; fi
if [ ! -e $PATH_To_AlignmentResults/Q20 ]; then mkdir $PATH_To_AlignmentResults/Q20 ; fi
if [ ! -e $PATH_To_AlignmentResults/Q20_rmdup ]; then mkdir $PATH_To_AlignmentResults/Q20_rmdup ; fi
if [ ! -e $PATH_To_AlignmentResults/bam2bed ]; then mkdir $PATH_To_AlignmentResults/bam2bed ; fi	

if [ ! -e $PATH_To_BAM ]; then mkdir $PATH_To_BAM ; fi
if [ ! -e $PATH_To_stats ]; then mkdir $PATH_To_stats ; fi
if [ ! -e $PATH_To_counts ]; then mkdir $PATH_To_counts ; fi


## Step 1, aligned the paired-end reads to the reference genome (here, mouse mm9) using BWA (v0.6.2). BWA should have been installed and you need to use full path of BWA or copy its executable to a directory which is in your shell's PATH.
# If the index has been done before, you need to construct the FM-index for the reference genome. For mouse mm9 genome, the command is below: 
bwa index -a bwtsw $PATH_To_Ref/mm9.fasta

# do the alignment (Here we just take one sample "s_7_TBL3S5" as an example. Basically, you need to do all samples one by one manually or using for loop to run all samples automaticaly)
bwa aln -t 16 $PATH_To_Ref/mm9.fasta $PATH_To_Fastq/s_7_TBL3S5_read1.fastq.gz > $PATH_To_AlignmentResults/s_7_TBL3S5_read1.sai
bwa aln -t 16 $PATH_To_Ref/mm9.fasta $PATH_To_Fastq/s_7_TBL3S5_read2.fastq.gz > $PATH_To_AlignmentResults/s_7_TBL3S5_read2.sai
bwa sampe -r '@RG\tID:MeDIP\tSM:s_7_TBL3S5\tPL:Illumina\tLB:QBI_library' $PATH_To_Ref/mm9.fasta $PATH_To_AlignmentResults/s_7_TBL3S5_read1.sai $PATH_To_AlignmentResults/s_7_TBL3S5_read2.sai $PATH_To_Fastq/s_7_TBL3S5_read1.fastq.gz $PATH_To_Fastq/s_7_TBL3S5_read2.fastq.gz > $PATH_To_AlignmentResults/s_7_TBL3S5.sam

# After finish alignment, we can generate the mapping statistics using the following PERL script. 
# "mm9_chr_name.txt" contains chromosome names that should be the same as that of the reference genome (eg. chr1, chrM, etc.), each chromosome per line. "*.sam" file is the input file generated by BWA. 
# "*_bwa_mappping_stats.xls" is the output file to list the summary of total number of reads, properly paired-end mapped reads, Properly paired-end mapped reads with good mapping quality and mapping stats based on each chromosome.
Get_properly_mapping_stats.pl $PATH_To_Ref/mm9_chr_name.txt $PATH_To_AlignmentResults/s_7_TBL3S5.sam $PATH_To_stats/s_7_TBL3S5_bwa_mappping_stats.xls

# ... please run all samples for bwa and "Get_properly_mapping_stats.pl" before go to next step (run Summary_stats.pl)

# After generate "_bwa_mappping_stats.xls" files for all samples, then run Summary_stats.pl to get the summary of mapping statistics for all samples
Summary_stats.pl $PATH_To_Ref/mm9.fa.fai $PATH_To_stats $PATH_To_AlignmentResults/bwa_map_stat.xls $PATH_To_AlignmentResults/bwa_mapchr_stats.xls $PATH_To_AlignmentResults/bwa_mapchr_stats_norm.xls

## Step 2, sort and index the “.bam” files, and remove duplicate reads using Samtools (v0.1.17). Samtools should have been installed and you need to use full path of BWA or copy its executable to a directory which is in your shell's PATH.
# sort the sam file and convert sam to bam 
samtools view -bt $PATH_To_Ref/mm9.fasta.fai $PATH_To_AlignmentResults/s_7_TBL3S5.sam | samtools sort - $PATH_To_AlignmentResults/s_7_TBL3S5.sort
# index the bam file
samtools index $PATH_To_AlignmentResults/s_7_TBL3S5.sort.bam

# select properly paired-end aligned reads with good mapping quality
samtools view -f 2 -h $PATH_To_AlignmentResults/s_7_TBL3S5.sort.bam | Map_Q20_sam.pl - $PATH_To_AlignmentResults/Q20/s_7_TBL3S5.sort.sam
samtools view -bt $ref $PATH_To_AlignmentResults/Q20/s_7_TBL3S5.sort.sam | samtools sort - $PATH_To_AlignmentResults/Q20/s_7_TBL3S5.sort

# remove duplicates
samtools rmdup $PATH_To_AlignmentResults/Q20/s_7_TBL3S5.sort.bam $PATH_To_AlignmentResults/Q20_rmdup/s_7_TBL3S5.sort.bam
# index the bam file
samtools index $PATH_To_AlignmentResults/Q20_rmdup/s_7_TBL3S5.sort.bam

# generate bed file based on the bam file (only used for properly paired-end aligned reads)
samtools view $PATH_To_AlignmentResults/Q20_rmdup/s_7_TBL3S5.sort.bam | sam2bed_PE.pl - $PATH_To_AlignmentResults/bam2bed/s_7_TBL3S5.bed


## Step 3, call peaks using MACS (v1.4.2)  
# Because macs14's behavior towards duplicate tags is not tuned for paired-end reads (It treats reads with same start coordination and the same strand as duplicates, which is ok for single-end reads but is not suitable for paired-end reads), we removed duplicates using samtools and using "--keep-dup=all" here.
macs14 -t $PATH_To_AlignmentResults/Q20_rmdup/s_7_TBL3S5.sort.bam -f BAM --keep-dup=all -g mm -p 1e-2 --nomodel --shiftsize 100 -n $PATH_To_peaks/s_7_TBL3S5 -B -S

# ... please run all samples for step 3 before go to step 4

## Step 4, all peak summits identified by MACS were then collected to get a full list of potential methylation sites. 
# All samples start with "s_7" or "s_8" will be collected in this example, eg. s_7_TBL3S1, s_7_TBL3S2, s_7_TBL3S3, s_7_TBL3S4, s_7_TBL3S5, s_8_TBL4S1, s_8_TBL4S2, etc.
MACS_1GetPeakSummits.pl $PATH_To_results/summits_position.xls $PATH_To_peaks s_7 s_8

## Step 5, generate raw count data and normalised count data
# Generate the count data and normalised count data for each sample or for multiple samples. It will save much computational time if the count data is generated based on each sample like following commands (eg., for sample "s_7_TBL3S5"), 
MACS_2ReadCount_BinarySearch.pl F $PATH_To_AlignmentResults/bwa_map_stat.xls $PATH_To_results/summits_position.xls $PATH_To_AlignmentResults/bam2bed/ $PATH_To_counts/s_7_TBL3S5.count.xls $PATH_To_counts/s_7_TBL3S5.count.norm.xls s_7_TBL3S5

# ... please run all samples for step 5 before go to step 6

## Step 6, applying Student's t-test to compare the count data between two groups
## For example, if group 1 has 7 biological replicates including "s_7_TBL3S1", "s_7_TBL3S2", "s_7_TBL3S3", "s_7_TBL3S4", "s_7_TBL3S5", "s_7_TBL3S6" and "s_7_TBL3S7", 
## and group 2 has 5 biological replicates including "s_8_TBL4S1", "s_8_TBL4S2", "s_8_TBL4S3", "s_8_TBL4S4" and "s_8_TBL4S5", the command will be like following:
MACS_3TTest.pl .count.norm.xls $PATH_To_counts/s_7_TBL3S1:$PATH_To_counts/s_7_TBL3S2:$PATH_To_counts/s_7_TBL3S3:$PATH_To_counts/s_7_TBL3S4:$PATH_To_counts/s_7_TBL3S5:$PATH_To_counts/s_7_TBL3S6:$PATH_To_counts/s_7_TBL3S7 $PATH_To_counts/s_8_TBL4S1:$PATH_To_counts/s_8_TBL4S2:$PATH_To_counts/s_8_TBL4S3:$PATH_To_counts/s_8_TBL4S4:$PATH_To_counts/s_8_TBL4S5 $PATH_To_results/results_pvalue.xls

## Step 7, select differnetially metylated sites based on P-value (say, p_value cutoff 0.05)
awk '$1=="Chr" || $NF<0.05' $PATH_To_results/results_pvalue.xls > $PATH_To_results/results_pvalue_DMS.xls

## Step 8, peak summits locate within 600bp were grouped together
MACS_4group_peaks.pl 600 $PATH_To_results/results_pvalue_DMS.xls $PATH_To_results/results_pvalue_DMS.group.xls
